### PR TITLE
feat(Threads): max autoArchiveDuration option

### PR DESCRIPTION
**Pull request by papsavas**
**Please describe the changes this PR makes and why it should be merged:**
This adds a `"MAX"` option to `ThreadAutoArchiveDuration` that sets a thread's auto-archive duration to the maximum one, based on the guilds boost count.

I understand that this is not really needed, so here's a couple of reasons this may or should get rejected:
- Runtime check
- Changes `ThreadAutoArchiveDuration` type to `number | string` (could be solved by setting all options to string and converting later) 
- Can be defined from moderators using channel default archived duration

**Status and versioning classification:**
- Code changes have been tested against the Discord API, or there are no code changes
- I know how to update typings and have done so, or typings don't need updating
- This PR changes the library's interface (methods or parameters added)

### CODE REVIEW

File: src/managers/ThreadManager.js [hunk 1]
```diff
@@ -66,7 +66,8 @@ class ThreadManager extends CachedManager {
    * * `1440` (1 day)
    * * `4320` (3 days) <warn>This is only available when the guild has the `THREE_DAY_THREAD_ARCHIVE` feature.</warn>
    * * `10080` (7 days) <warn>This is only available when the guild has the `SEVEN_DAY_THREAD_ARCHIVE` feature.</warn>
-   * @typedef {number} ThreadAutoArchiveDuration
+   * * `'MAX'` Based on the guilds boost count
+   * @typedef {number | string} ThreadAutoArchiveDuration
```
Comment (NotSugden) [inline src/managers/ThreadManager.js:70]:
```suggestion
   * @typedef {number|string} ThreadAutoArchiveDuration
```

File: src/managers/ThreadManager.js [hunk 2]
```diff
@@ -119,6 +120,15 @@ class ThreadManager extends CachedManager {
     } else if (this.channel.type !== 'GUILD_NEWS') {
       resolvedType = typeof type === 'string' ? ChannelTypes[type] : type ?? resolvedType;
     }
+    if (autoArchiveDuration === 'MAX') {
+      let maxAutoArchiveDuration = 1440;
```
Comment (NotSugden) [inline src/managers/ThreadManager.js:124]: why not just scrap thia variable and assign to `autoArchiveDuration` directly

**Review by NotSugden — CHANGES_REQUESTED** [review]

File: src/structures/ThreadChannel.js [hunk 3]
```diff
@@ -319,6 +327,14 @@ class ThreadChannel extends Channel {
...
    */
   setAutoArchiveDuration(autoArchiveDuration, reason) {
+    if (autoArchiveDuration === 'MAX') {
+      autoArchiveDuration = 1440;
+      if (this.guild.premiumSubscriptionCount >= 15) {
+        autoArchiveDuration = 10080;
+      } else if (this.guild.premiumSubscriptionCount >= 2) {
+        autoArchiveDuration = 4320;
+      }
+    }
```
Comment (ckohen) [inline src/structures/ThreadChannel.js:337]: This is unnecessary, edit already handles it

File: src/structures/ThreadChannel.js [hunk 4]
```diff
@@ -275,6 +275,14 @@ class ThreadChannel extends Channel {
    *   .catch(console.error);
    */
   async edit(data, reason) {
+    if (data.autoArchiveDuration === 'MAX') {
+      data.autoArchiveDuration = 1440;
+      if (this.guild.premiumSubscriptionCount >= 15) {
+        data.autoArchiveDuration = 10080;
+      } else if (this.guild.premiumSubscriptionCount >= 2) {
+        data.autoArchiveDuration = 4320;
```
Comment (ckohen) [inline src/structures/ThreadChannel.js:283]: Check the guild features, not the subscription count

File: src/structures/Message.js [hunk 5]
```diff
@@ -763,6 +763,14 @@ class Message extends Base {
...
     }
     if (this.hasThread) return Promise.reject(new Error('MESSAGE_EXISTING_THREAD'));
+    if (options.autoArchiveDuration === 'MAX') {
+      options.autoArchiveDuration = 1440;
+      if (this.guild.premiumSubscriptionCount >= 15) {
+        options.autoArchiveDuration = 10080;
+      } else if (this.guild.premiumSubscriptionCount >= 2) {
+        options.autoArchiveDuration = 4320;
+      }
+    }
```
Comment (ckohen) [inline src/structures/Message.js:773]: ThreadManager#create already does this

File: src/managers/ThreadManager.js [hunk 6]
```diff
@@ -119,6 +120,14 @@ class ThreadManager extends CachedManager {
...
       resolvedType = typeof type === 'string' ? ChannelTypes[type] : type ?? resolvedType;
     }
+    if (autoArchiveDuration === 'MAX') {
+      autoArchiveDuration = 1440;
+      if (this.channel.guild.premiumSubscriptionCount >= 15) {
+        autoArchiveDuration = 10080;
+      } else if (this.channel.guild.premiumSubscriptionCount >= 2) {
+        autoArchiveDuration = 4320;
+      }
+    }
```
Comment (ckohen) [inline src/managers/ThreadManager.js:130]: Use guild features

**Review by ckohen — CHANGES_REQUESTED** [review]

File: src/structures/ThreadChannel.js [hunk 7]
```diff
@@ -277,9 +277,9 @@ class ThreadChannel extends Channel {
   async edit(data, reason) {
     if (data.autoArchiveDuration === 'MAX') {
       data.autoArchiveDuration = 1440;
-      if (this.guild.premiumSubscriptionCount >= 15) {
+      if (this.channel.guild.features.includes('SEVEN_DAY_THREAD_ARCHIVE')) {
```
Comment (iCrawl) [inline src/structures/ThreadChannel.js:280]: Why access the channel here?

File: src/structures/ThreadChannel.js [hunk 8]
```diff
@@ -277,9 +277,9 @@ class ThreadChannel extends Channel {
...
         data.autoArchiveDuration = 10080;
-      } else if (this.guild.premiumSubscriptionCount >= 2) {
+      } else if (this.channel.guild.features.includes('THREE_DAY_THREAD_ARCHIVE')) {
```
Comment (iCrawl) [inline src/structures/ThreadChannel.js:282]: Ditto

Comment (papsavas) [reply src/structures/ThreadChannel.js:282]: haha yeah, fixing rn, missed it in copypaste

**Review by iCrawl — APPROVED** [review]

**Review by kyranet — APPROVED** [review]
Maybe we can make a new method in Util to fetch the highest duration available for a guild, but otherwise this is fine.

File: src/structures/ThreadChannel.js [hunk 9]
```diff
@@ -275,6 +275,14 @@ class ThreadChannel extends Channel {
...
    */
   async edit(data, reason) {
+    if (data.autoArchiveDuration === 'MAX') {
+      data.autoArchiveDuration = 1440;
+      if (this.guild.features.includes('SEVEN_DAY_THREAD_ARCHIVE')) {
+        data.autoArchiveDuration = 10080;
+      } else if (this.guild.features.includes('THREE_DAY_THREAD_ARCHIVE')) {
+        data.autoArchiveDuration = 4320;
+      }
+    }
```
Comment (vladfrangu) [inline src/structures/ThreadChannel.js:285]: Any way this could not mutate the options?

**Review by vladfrangu — DISMISSED** [review]

File: src/structures/ThreadChannel.js [hunk 10]
```diff
@@ -275,11 +275,20 @@ class ThreadChannel extends Channel {
    *   .catch(console.error);
    */
   async edit(data, reason) {
+    let maxAutoArchiveDuration;
```
Comment (NotSugden) [inline src/structures/ThreadChannel.js:278]:
```suggestion
    let maxAutoArchiveDuration = data.autoArchiveDuration;
```
and remove the `?? ...` at the end

**Review by vladfrangu — APPROVED** [review]
I'd really love to see this filtering get moved to a util function, but LGTM I suppose

**Review by SpaceEEC — APPROVED** [review]

**Pull request merged**